### PR TITLE
add a new flag --ignore-not-found to vclusterctl

### LINKED ISSUE #771: `vcluster delete --ignore-not-found` just like `kubectl`
### Is your feature request related to a problem?

I'm using vcluster in CI, and I add `vcluster delete` as a cleanup step of my pipeline. Sometimes the pipeline fails before the vcluster gets created, and therefore `vcluster delete` fails with `couldn't find vcluster`.

### Which solution do you suggest?

A new flag like `vcluster delete --ignore-not-found`, similar to `kubectl delete --ignore-not-found`. It would prevent returning a non-zero status code in case the vcluster does not exist already. Also, it would still attempt to delete other resources if told to, like `--delete-namespace` and the PVC, but also would not fail if the PVC or Namespace is not found.

### Which alternative solutions exist?

A wrapper for `vcluster delete`:

```
#!/bin/bash

set -eux

code=0

# Ignore when vcluster isn't found
if ! output=$(vcluster delete vcluster --namespace my-vcluster 2>&1) &&
    [[ "${output}" != *"couldn't find vcluster"* ]]; then
    code=1
fi
echo "${output}"

kubectl delete namespace my-vcluster --ignore-not-found || code=1

exit "${code}"
```

### Additional context

*No response*

### PR DESCRIPTION
**Pull request by rahulii**
**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind feature

**What does this pull request do? Which issues does it resolve?** (use `resolves #<issue_number>` if possible) 
resolves #771 

**Please provide a short message that should be published in the vcluster release notes**
Added a new flag `--ignore-not-found` that will not error out in case the vcluster is not found while deleting it

**What else do we need to know?**

### CODE REVIEW

**Comment by FabianKramm** [comment]
@rahulii thanks for the PR! Looks like you need to rebase the PR on the current master, would be great if you could do that!

**Comment by rahulii** [comment]
@FabianKramm @matskiv @mahendrabagul  , if the vcluster is not found (it returns deom here https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/delete.go#L160). Here I check if --ignore-not-found flag is set and the error is "Not Found error" to not error out. But from here,  In case user says `vcluster delete my-cluster --namespace my-namespace --delete-namespace --ignore-not-found` the kubeClient is nil , and hence it panics. 
My question is: in case --ignore-not-found is set and vcluster is not found, do we construct a kubeclient from current context  using https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/find/find.go#L42?

**Comment by matskiv** [comment]
@rahulii I think you will need to modify [the prepare function](https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/delete.go#L157) to initialize the cmd.kubeClient before it calls find.GetVCluster

File: cmd/vclusterctl/cmd/delete.go [hunk 1]
```diff
@@ -144,7 +156,11 @@ func (cmd *DeleteCmd) Run(cobraCmd *cobra.Command, args []string) error {
 		err = client.CoreV1().Namespaces().Delete(context.Background(), cmd.Namespace, metav1.DeleteOptions{})
 		if err != nil {
 			if !kerrors.IsNotFound(err) {
-				return errors.Wrap(err, "delete namespace")
+				if cmd.IgnoreNotFound{
```
Comment (matskiv) [inline cmd/vclusterctl/cmd/delete.go:159]: Same as above - I don't think this is necessary as the line above ignores the "NotFound" error.

**Comment by rahulii** [comment]
> @rahulii I think you will need to modify [the prepare function](https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/delete.go#L157) to initialize the cmd.kubeClient before it calls find.GetVCluster

@matskiv not sure how is that possible? here the [restConfig](https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/delete.go#L174) is fetched using vCluster.ClientFactory.ClientConfig().

**Comment by rahulii** [comment]
> @rahulii I think you will need to modify [the prepare function](https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/delete.go#L157) to initialize the cmd.kubeClient before it calls find.GetVCluster

@matskiv here is the approach I was thinking:
func (cmd *DeleteCmd) prepare(vClusterName string) error {
if cmd.IgnoreNotFound {
   context := CurrentContext()
   // construct kubeclient using this context
} else {
  // current approach 
}

}

**Comment by matskiv** [comment]
@rahulii I think you don't even need to check the flag in the prepare function. Just move this part to the very top of the prepare function, before GetVCluster is called:
```
kubeClient, err := kubernetes.NewForConfig(restConfig)
if err != nil {
  return err
}
cmd.kubeClient = kubeClient
```

**Comment by rahulii** [comment]
> restConfig

**Comment by rahulii** [comment]
> @rahulii I think you don't even need to check the flag in the prepare function. Just move this part to the very top of the prepare function, before GetVCluster is called:
> 
> ```
> kubeClient, err := kubernetes.NewForConfig(restConfig)
> if err != nil {
>   return err
> }
> cmd.kubeClient = kubeClient
> ```

@matskiv umm to create a kubeClient it needs a restConfig which is derived as:
```restConfig, err := vCluster.ClientFactory.ClientConfig()
	if err != nil {
		return err
	}```
here - https://github.com/loft-sh/vcluster/blob/main/cmd/vclusterctl/cmd/delete.go#L178

**Comment by matskiv** [comment]
@rahulii yes, good point, I overlook that. But I think we can still do it, we just need to provide the restConfig like so:
```
restConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(clientcmd.NewDefaultClientConfigLoadingRules(), &clientcmd.ConfigOverrides{}).ClientConfig()
```
that^ is pretty much equivalent to the `vCluster.ClientFactory.ClientConfig()`.

**Comment by matskiv** [comment]
Hello @rahulii 
Do you plan to finish this PR or should we unassign you from the issue and close the PR for now? If we close, are you okay with someone else cherry-picking your changes into a new PR and adding the remaining code?

**Comment by rahulsawra98** [comment]
> Hello @rahulii Do you plan to finish this PR or should we unassign you from the issue and close the PR for now? If we close, are you okay with someone else cherry-picking your changes into a new PR and adding the remaining code?

hey @matskiv I will be completing it very soon. Apologies for delay!

**Comment by rahulii** [comment]
> Hello @rahulii Do you plan to finish this PR or should we unassign you from the issue and close the PR for now? If we close, are you okay with someone else cherry-picking your changes into a new PR and adding the remaining code?

Hey @matskiv I am on it. Really sorry for delays!

**Comment by heiko-braun** [comment]
Outdated

**Pull request closed without merging**